src/Aspire.Hosting.Keycloak/KeycloakResource.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Aspire.Hosting.ApplicationModel;
+
+/// <summary>
+/// A resource that represents a Keycloak resource.
+/// </summary>
+public sealed class KeycloakResource : ContainerResource, IResourceWithServiceDiscovery
+{
+    private const string DefaultAdmin = "admin";
+    internal const string PrimaryEndpointName = "tcp";
+
+    /// <param name="name">The name of the resource.</param>
+    /// <param name="admin">A parameter that contains the Keycloak admin, or <see langword="null"/> to use a default value.</param>
+    /// <param name="adminPassword">A parameter that contains the Keycloak admin password.</param>
+    public KeycloakResource(string name, ParameterResource? admin, ParameterResource adminPassword) : base(name)
+    {
+        ArgumentNullException.ThrowIfNull(adminPassword);
+
+        AdminParameter = admin;
+        AdminPasswordParameter = adminPassword;
+    }
+
+    private EndpointReference? _primaryEndpoint;
+
+    /// <summary>
+    /// Gets the parameter that contains the Keycloak admin.
+    /// </summary>
+    public ParameterResource? AdminParameter { get; }
+
+    internal ReferenceExpression AdminReference =>
```

Comment:
I think to make the resource type cleaner we should take care of setting up the defaults inside the extension methods so that we just have two properties `UserNameParameter` and `PasswordParameter`.

Reply:
> I think to make the resource type cleaner we should take care of setting up the defaults inside the extension methods so that we just have two properties `UserNameParameter` and `PasswordParameter`.

Or `AdminUserNameParameter` and `AdminPasswordParameter`.

Reply:
That works too. Given this context `UserNameParameter` might not be sufficient clarity.

Reply:
I see a few other resources (RabbitMQ and Postgress) use the same pattern I used here. 

Happy to make the change, but would you point me to an example to follow?